From ef69918c044a527e154a919c829076bc19689925 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Thu, 29 Aug 2013 22:25:14 +0200
Subject: wined3d: Don't lock the src volume in device_update_volume

This will attempt to wait for the CS, but this code is executed by the CS, so
it will, among other problems, break the single producer, single consumer
assumption of the work queue.

FIXME: Maybe merge this with the previous patch or change their order.

FIXME 2: Loading the src location might be nice, or even better,
responding to unexpected src locs. In theory it should be in SYSMEM
since its a sysmem texture.
---
 dlls/wined3d/device.c | 15 ++++-----------
 1 file changed, 4 insertions(+), 11 deletions(-)

diff --git a/dlls/wined3d/device.c b/dlls/wined3d/device.c
index 955f061..12e87e1 100644
--- a/dlls/wined3d/device.c
+++ b/dlls/wined3d/device.c
@@ -3496,8 +3496,7 @@ static void wined3d_device_update_texture_3d(struct wined3d_context *context,
         struct wined3d_texture *src_texture, unsigned int src_level,
         struct wined3d_texture *dst_texture, unsigned int level_count)
 {
-    struct wined3d_const_bo_address data;
-    struct wined3d_map_desc src;
+    struct wined3d_bo_address data;
     unsigned int i;
 
     TRACE("context %p, src_texture %p, src_level %u, dst_texture %p, level_count %u.\n",
@@ -3509,17 +3508,11 @@ static void wined3d_device_update_texture_3d(struct wined3d_context *context,
 
     for (i = 0; i < level_count; ++i)
     {
-        if (FAILED(wined3d_resource_map(&src_texture->resource,
-                src_level + i, &src, NULL, WINED3D_MAP_READONLY)))
-            return;
+        wined3d_texture_get_memory(src_texture, src_level + i, &data,
+                src_texture->resource.map_binding);
 
-        data.buffer_object = 0;
-        data.addr = src.data;
-        wined3d_volume_upload_data(dst_texture, i, context, &data);
+        wined3d_volume_upload_data(dst_texture, i, context, wined3d_const_bo_address(&data));
         wined3d_texture_invalidate_location(dst_texture, i, ~WINED3D_LOCATION_TEXTURE_RGB);
-
-        if (FAILED(wined3d_resource_unmap(&src_texture->resource, src_level + i)))
-            return;
     }
 }
 
-- 
2.8.0

